spec/models/form_spec.rb

```diff
@@ -296,4 +296,44 @@
       end
     end
   end
+
+  describe "#move_to_group" do
+    let(:form) { create(:form_record) }
+    let(:old_group) { create :group, name: "Group 1" }
+
+    before do
+      GroupForm.create!(form_id: form.id, group: old_group)
+    end
+
+    context "when the form is in an existing group" do
+      context "when the user tries to move the form to a different group" do
+        let(:new_group) { create :group, name: "Group 2" }
+        let(:group_form) { GroupForm.find_by!(form_id: form.id) }
+
+        it "moves the form" do
+          expect { form.move_to_group(new_group.external_id) }.to change { GroupForm.find_by!(form_id: form.id).reload.group.id }.to(new_group.id)
+        end
+      end
+
+      context "when the user tries to move the form to its existing group" do
+        it "leaves the form where it is" do
+          group_form = GroupForm.find_by!(form_id: form.id)
+
+          expect { form.move_to_group(old_group.external_id) }.not_to(change { group_form.reload.group.id })
+        end
+      end
+
+      context "when the new group does not exist" do
+        it "throws an ActiveRecord::RecordNotFound error" do
+          expect { form.move_to_group("some_nonexistent_external_id") }.to raise_error(ActiveRecord::RecordNotFound)
+        end
+      end
+    end
+
+    context "when the form is not already in a group" do
+      it "throws an ActiveRecord::RecordNotFound error" do
+        expect { form.move_to_group("some_nonexistent_external_id") }.to raise_error(ActiveRecord::RecordNotFound)
```

Comment:
Is it possible for a form to not be in a group?

Reply:
I think the before block needs to be inside the other context for this to not have a group? Or at least not have a GroupForm...

Reply:
Good spot, have moved the before block!

It might be overly defensive - there _shouldn't_ be any forms in that state since we moved them all in bulk when we added groups. I figured it might still be theoretically possible, so if it happens an error would be useful so we know about it and can figure out why it happened. But I'm happy to remove it if you reckon it's unnecessary!